functions.ipynb

```diff
@@ -0,0 +1,1490 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "b6e0ee24-8115-4741-acfe-c2938bb40525",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "%%ipytest\n",
+    "\n",
+    "def solution_longest_sequence_best(nums: \"list[int]\") -> int:\n",
+    "    \"\"\"\n",
+    "    Write your solution here\n",
+    "    \"\"\"\n",
+    "    pass"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "1c93c572-7890-4bd7-a0f6-d4015679df39",
+   "metadata": {
+    "jp-MarkdownHeadingCollapsed": true,
+    "tags": []
+   },
+   "source": [
+    "## Password validator"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6a40b25b-fa38-44c8-8a80-d1e86c38ecab",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "### Part 1 🌶️"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "61fdaa6e-9f5c-4039-a09e-e79146175703",
+   "metadata": {},
+   "source": [
+    "You have a range of numbers `136760-595730` and need to count how many valid password it contains. A valid password must meet **all** the following criteria:\n",
+    "\n",
+    "- It is a six-digit number\n",
+    "- Two adjacent digits are the same (like `22` in `122345`)\n",
+    "- Going from left to right, the digits **never decrease**; they only ever increase or stay the same (like 111123 or 135679)\n",
+    "\n",
+    "\n",
+    "---\n",
+    "\n",
+    "\n",
+    "For example, the following are true:\n",
+    "\n",
+    "- `111111` meets these criteria (double `11`, never decreases)\n",
+    "- `223450` does **not** meet these criteria (`50` is a decreasing pair of digits)\n",
+    "- `123789` does **not** meet these criteria (no double digit)\n",
+    "\n",
+    "Write a function that determines if a password is valid according to the criteria above.\n",
+    "\n",
+    "**How many valid password are there in your range?**\n",
+    "\n",
+    "*Hint:* the solution function should take **two** arguments: the `start` and `end` of your password range"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "517f4d43-0ffa-407d-9035-7c8e47b64103",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%ipytest\n",
+    "\n",
+    "def solution_password_validator1(start: int, end: int) -> int:\n",
+    "    \"\"\"\n",
+    "    Write your solution here\n",
+    "    \"\"\"\n",
+    "    pass"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "edac5c6a-23fa-4f09-a462-5682201a449f",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "---\n",
+    "\n",
+    "### Part 2 🌶️🌶️"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "af330663-15c2-48f7-b407-e09cecc62d4d",
+   "metadata": {},
+   "source": [
+    "You have a new rule: the two adjacent matching digits **must not be part of a larger group of matching digits**.\n",
```

Comment:
This would break the solution because `111122` is a valid password, but if we enforce to have **at max** two matching digits, that password isn't valid anymore.

I could use a better rephrasing like:

> You must not have a larger group of identical digits that include the two adjacent matching digits

Reply:
I thought `the two adjacent matching digits must not be part of a larger group of matching digits` means `111`, `222`, and `333` are all forbidden. 

Because two adjacent matching digits (`11`, `22`, `33`) are part of a larger group(`111`, `222`, `333`).

Sorry, I am totally confused 😮

Reply:
It can be indeed confusing, I agree. I think that the following has no more ambiguity:

> **at least** two adjacent matching digits must not be part of a larger group of matching digits